### PR TITLE
MM-52754: Publish new schema migration guidelines

### PR DESCRIPTION
**Pull request by agnivade**
https://mattermost.atlassian.net/browse/MM-52754

I have re-purposed the old page to house the latest documentation that we want to publish: https://docs.google.com/document/d/1wdQLzS_C6cjJIcyZWEYEeaP3Ej_NuuXj4lDt1VFuyZU/edit#heading=h.d8w09w4neiea.

Perhaps we can still keep the morph commands. Let me know if that's useful.

### CODE REVIEW

**Comment by streamer45** [comment]
> Perhaps we can still keep the morph commands. Let me know if that's useful.

Anything that's still applicable today should be retained in my opinion. Whether you want to move it somewhere else or restructure it up to you.

**Comment by agnivade** [comment]
Let me restructure.

**Comment by agnivade** [comment]
Done

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 1]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+| DROP COLUMN    | YES           | YES                    |
+| ADD FK CONSTRAINT| NO          | NO (avoidable)<sup>3</sup>        |
+| ADD UNIQUE CONSTRAINT | NO     | YES                    |
+
+Note:
+1. Fulltext and spatial indexes take locks. Adding a fulltext index can also rewrite the table if there is no user-defined FTS_DOC_ID column, or in other words if a fulltext index is being added for the first time.
+2. Concurrent DML is also not permitted while adding an auto-increment column. However we don’t use it and are unlikely to use it in future. So I have made the cell green for now.
+3. Setting `foreign_key_checks` to false avoids table locks. But it also defeats the purpose of a foreign key.
+
+**MySQL 8**
```
Comment (isacikgoz) [inline site/content/contribute/more-info/server/schema-migration-guide.md:82]: On a editorial note, shall we use a single table for mysql? The only difference is at adding column we can add the exception for that in the footnote like others.

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 2]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+| ADD COLUMN     | NO (only in certain cases) <sup>2</sup>           | YES<sup>3</sup>                   |
+| ALTER COLUMN   | YES           | NO                     |
+| DROP COLUMN    | YES           | YES                    |
+| ADD FK CONSTRAINT| NO          | NO (avoidable)<sup>4</sup>        |
+| ADD UNIQUE CONSTRAINT | NO     | YES                    |
+
+Note:
+1. Fulltext and spatial indexes take locks. Adding a fulltext index can also rewrite the table if there is no user-defined FTS_DOC_ID column, or in other words if a fulltext index is being added for the first time.
+2. Table rewrite happens if the column is being added to a table with a fulltext index. Which means this will affect adding columns to the Posts or the Channels table for example.
+3. Concurrent DML is also not permitted while adding an auto-increment column. However we don’t use it and are unlikely to use it in future. So I have made the cell green for now.
```
Comment (isacikgoz) [inline site/content/contribute/more-info/server/schema-migration-guide.md:97]: I think markdown wouldn't allow for colored tables?

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 3]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+
+**Postgres**: Relatively rare, but out of those 6 cases, 2 are adding unique constraints. For example:
+
+```sql
+ALTER TABLE oauthaccessdata ADD CONSTRAINT oauthaccessdata_clientid_userid_key UNIQUE (clientid, userid);
+```
+
+This can be improved by first adding the index concurrently, and then attaching the index to the constraint. See example later.
+
+Adding a foreign key in postgres takes a share row exclusive lock, which means only SELECT queries are allowed. It is possible to bypass the table scanning by adding a “NOT VALID” suffix, but then it defeats the purpose of having a FK. Suggestion would be to avoid doing it.
```
Comment (cwarnermm) [inline site/content/contribute/more-info/server/schema-migration-guide.md:165]:
```suggestion
Adding a foreign key in PostgreSQL takes a share row exclusive lock, which means only SELECT queries are allowed. It is possible to bypass the table scanning by adding a “NOT VALID” suffix, but then it defeats the purpose of having a foreign key. We recommend against doing it.
```

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 4]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+
+## Examples
+
+### How do I change a column type if I MUST
+
+Follow this long-winded procedure:
+- Create a new column.
+- Migrate existing data.
+- From next ESR, start using the new column.
+- Next ESR, drop the old column.
```
Comment (isacikgoz) [inline site/content/contribute/more-info/server/schema-migration-guide.md:229]: What do you think about helping readers by adding a visual for this? Maybe it can be easier to understand.

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 5]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
-The problem arises when in some databases, for some tables, due to various technical reasons, applying a migration prevents other operations from happening on that table. In that case, it causes unavoidable downtime.
+Adding an FK takes locks. But avoidable by setting `SET  foreign_key_checks=0`. Then it doesn't take any locks, and neither rebuilds the table. But it partially defeats the purpose of using an FK. Recommend to AVOID.
 
-### TL;DR
+6. DROP INDEX
 
-1. Migrations are always synchronous.
-2. The time taken to run a migration is not always the primary factor in judging impact.
-3. Except in cases where a migration cannot run concurrently with other database operations.
+**Postgres**: DROP INDEX CONCURRENTLY does not take any locks.
```
Comment (cwarnermm) [inline site/content/contribute/more-info/server/schema-migration-guide.md:175]:
```suggestion
**PostgreSQL**: DROP INDEX CONCURRENTLY does not take any locks.
```

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 6]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+
+3. Now we can start using the new column from the next ESR version. But we cannot yet drop the existing column because of backwards compatibility guarantees. The old column would still be in use by older app nodes in the cluster during upgrade. We also would want to drop the trigger since its use is finished.
+
+```sql
+DROP TRIGGER tr_update_status_channel_count on status
+```
+
+4. And in the ESR after that, now we can finally drop the old column.
+
+This deliberately skips renaming the column for simplicity. Depending on your use-case, you can do that if you want to. It is a fast operation that does not rebuild the table in both MySQL and Postgres, so there are no issues.
```
Comment (cwarnermm) [inline site/content/contribute/more-info/server/schema-migration-guide.md:282]:
```suggestion
This deliberately skips renaming the column for simplicity. Depending on your use-case, you can do that if you want to. It is a fast operation that does not rebuild the table in both MySQL and PostgreSQL, so there are no issues.
```

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 7]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+
+### How do I add a unique constraint to a table (postgres only)
```
Comment (cwarnermm) [inline site/content/contribute/more-info/server/schema-migration-guide.md:284]:
```suggestion
### How do I add a unique constraint to a table (PostgreSQL only)
```

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 8]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+
+For example, let’s say the next upcoming version is 8.4, and the next ESR is 8.6. So step 1 and 2, goes in 8.4. And in 8.7 onwards, we add the code to start using the new column, which will eventually be part of 8.12 (ESR after that). And then from 8.13 onwards, we can drop the column.
+
+The reasoning behind this is some customers will only upgrade from ESR to ESR. So we need to ensure backwards compatibility with the previous version.
+
+Following shows an example where we are adding a channel_count column to the status table. This is not exactly altering a column, but the idea remains the same, and you can extend this to fit your use-case.
+
+1. ALTER TABLE status ADD COLUMN channel_count integer;
+
+2. Our next objective is to migrate existing data. We do this in a 2-phase approach where we set up triggers to migrate all new data and in the background migrate old data in batches.
```
Comment (isacikgoz) [inline site/content/contribute/more-info/server/schema-migration-guide.md:239]: How are we going to be sure the migration is completed with the trigger mechansim? I mean are we thinking of a mechanism to enable admisitrators to follow progress? Also can you elaborate how resilient is this mechanism?

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 9]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+This can be improved by first adding the index concurrently, and then attaching the index to the constraint. See example later.
+
+Adding a foreign key in postgres takes a share row exclusive lock, which means only SELECT queries are allowed. It is possible to bypass the table scanning by adding a “NOT VALID” suffix, but then it defeats the purpose of having a FK. Suggestion would be to avoid doing it.
+
+**MySQL**:
+
+Adding an unique constraint doesn't take locks, but does a table rebuild.
 
-The problem arises when in some databases, for some tables, due to various technical reasons, applying a migration prevents other operations from happening on that table. In that case, it causes unavoidable downtime.
+Adding an FK takes locks. But avoidable by setting `SET  foreign_key_checks=0`. Then it doesn't take any locks, and neither rebuilds the table. But it partially defeats the purpose of using an FK. Recommend to AVOID.
```
Comment (cwarnermm) [inline site/content/contribute/more-info/server/schema-migration-guide.md:171]:
```suggestion
Adding a foreign key takes locks. But avoidable by setting `SET  foreign_key_checks=0`. Then it doesn't take any locks, and neither rebuilds the table. But it partially defeats the purpose of using an foreign key. We recommend you avoid this.
```

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 10]
```diff
@@ -44,16 +339,9 @@ The problem arises when in some databases, for some tables, due to various techn
     - If it's applied you can revert it using morph CLI command. An example command would look like `morph apply down --driver {your-driver} --dsn "{your-dsn}" --path {path-to-your-driver-specific-migration-files} --number 1`
 2. If the migration has been shipped in a release and you want to apply fixes, instead of changing the existing script, you should add a new one so that `db_migrations` will stay consistent. You can edit the existing migration to be a no-op for future releases in this case.
 
-### How do I measure the impact of the migration?
```
Comment (isacikgoz) [inline site/content/contribute/more-info/server/schema-migration-guide.md:47]: I think we should still have some sort of imapct analysis of migration (for the obvious ones) 0/5

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 11]
```diff
@@ -44,16 +339,9 @@ The problem arises when in some databases, for some tables, due to various techn
...
-Answer the following two questions:
-1. Is it a large table (like `Posts` or `ThreadMemberships`)? (The performance team can help you with this.)
-2. Can my change run concurrently with other database operations in that table? Check MySQL and PostgreSQL documentation, starting from required minimum versions supported by Mattermost.
-
-If the answer to 2. is yes, then it should be pretty simple. In that case, if the answer to 1 is also yes, then please extract the raw SQL query to be run and add it in the release notes for customers who want to run it in off-hours to speed up upgrade time.
-
-However, If the answer to the second question above is "no", then contact the {{< newtabref href="https://community.mattermost.com/core/channels/developers-performance" title="performance team" >}} to discuss a solution.
+## GLOSSARY
 
-### Inform SRE team
```
Comment (isacikgoz) [inline site/content/contribute/more-info/server/schema-migration-guide.md:57]: I think we should still keep this recommendation, just in cae :)

**Review by isacikgoz — COMMENTED** [review]
Did my first pass, have some questions but this is a great guide! Good stuff 🚀

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 2 above)
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:97]: Yeah I looked very slightly into this. You _could_ technically do this in hugo with some custom css, but I didn't bother spending that much effort :stuck_out_tongue:

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 11 above)
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:57]: This was initially introduced by the SRE team but is not enforced anymore at the moment. And moreover, we would be stricly following non-blocking migrations and along with Community moving to Postgres, I don't think this is necessary anymore. LMK.

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 2 above)
Comment (cwarnermm) [reply site/content/contribute/more-info/server/schema-migration-guide.md:97]: If this is highly desirable, @emdecr may be able to assist.

Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:97]: It will be good to have ;)

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 10 above)
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:47]: The whole idea is actually covered by the earlier sections where we talk about how we avoid non-blocking migrations. So we don't really need to extract the raw SQL anymore. I am trying to keep things simple. But perhaps, we can keep a line on calculating schema migration time. I'll add that as a last step in the previous section.

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 8 above)
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:239]: The trigger will only be executed for any new rows or updates. So there is no completion marker there. The completion marker is for the job, which will just add a new row in the Systems table when done. The reliability of that is handled by the jobs infrastructure, which we have been using since the start. The trigger resides in the DB itself, so no issues on that reliability wise.

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 4 above)
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:229]: Sure, let me add a diagram

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 12]
```diff
@@ -7,6 +7,23 @@ aliases:
...
+This document aims to do an analysis of the types of schema migrations we do in Mattermost and ways to make them non-blocking so as to improve the Mattermost upgrade experience.
+Historically, we have never put a lot of thought to the migration process. Developers would simply add a DDL statement and call it a day. But that causes a significant impact to large customers for whom downtime is not an option. This causes them to push back on upgrading their Mattermost version for a long time (sometimes for several years). This in turn, has some cascading effects like customers not able to get new features, performance improvements etc.
+
+We want to improve the situation and make upgrades a worry-free experience for our customers. It definitely comes at a cost of writing more code and delaying some features to avoid breaking changes. This document will aim to uncover all those cases and provide best practices to follow so that we can hit the right balance.
+
+## Goals
+
+We have two overarching goals and a third auxiliary goal.
+
+1. Schema migrations should **ALWAYS** be backwards compatible till the last ESR.
```
[the remark this answers is not in the document]
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:21]: No, the last ESR from the version the migration is being introduced.

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 1 above)
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:82]: Sure

**Comment by agnivade** [comment]
@isacikgoz - PTAL

File: site/content/contribute/more-info/server/schema-migration-guide.md [hunk 13]
```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
...
+| DROP INDEX     | NO            | YES                    |
+| ADD COLUMN     | YES           | YES<sup>2</sup>                   |
+| ALTER COLUMN   | YES           | NO                     |
+| DROP COLUMN    | YES           | YES                    |
+| ADD FK CONSTRAINT| NO          | NO (avoidable)<sup>3</sup>        |
+| ADD UNIQUE CONSTRAINT | NO     | YES                    |
+
+Note:
+1. Fulltext and spatial indexes take locks. Adding a fulltext index can also rewrite the table if there is no user-defined FTS_DOC_ID column, or in other words if a fulltext index is being added for the first time.
+2. Concurrent DML is also not permitted while adding an auto-increment column. However we don’t use it and are unlikely to use it in future. So I have made the cell green for now.
```
[the remark this answers is not in the document]
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md]: I realized there is no "green" color here :) Since this was copied from the google doc. Removed the last sentence.

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 8 above)
Comment (isacikgoz) [reply site/content/contribute/more-info/server/schema-migration-guide.md:228]: Oh okay, I missed the job part being handled by the application. Thanks for the clarification 👍

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 10 above)
Comment (isacikgoz) [reply site/content/contribute/more-info/server/schema-migration-guide.md:47]: Yep, a single sentence would be great. We may still want some review/discussion mechanism on any migrations that would happening on large tables.

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 11 above)
Comment (isacikgoz) [reply site/content/contribute/more-info/server/schema-migration-guide.md:57]: Fair enough, let's remove this section.

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 10 above)
Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md:47]: Yep the line is added now. And yes, we have CODEOWNERS in the migrations directory now. So all and any migration will have to be reviewed by someone from our team. We will discuss any large migration on the PR itself.

**Review by isacikgoz — APPROVED** [review]
LGTM, awesome resource!

File: site/content/contribute/more-info/server/schema-migration-guide.md (hunk 2 above)
Comment (emdecr) [reply site/content/contribute/more-info/server/schema-migration-guide.md]: @agnivade @cwarnermm Sorry I missed this - my notification routing was weird; didn't get pinged on my MM email 😅 

Anywho - Did y'all still want me to take a look at this? Just lemme know. Screencap attached. Was it this entire column that you wanted to colour green?

Screen Shot 2023-05-16 at 10 37 29 AM

Thanks!

Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md]: @emdecr - Not the entire column, check this [doc](https://docs.google.com/document/d/1wdQLzS_C6cjJIcyZWEYEeaP3Ej_NuuXj4lDt1VFuyZU/edit#heading=h.xe5arfvo4667). Some cells will be green, red and yellow depending on their value. Basically we need to match whatever is there in the doc.

Let us know if it's doable :)

Comment (agnivade) [reply site/content/contribute/more-info/server/schema-migration-guide.md]: No rush at all. Feel free to take your time!

**Comment by emdecr** [comment]
@agnivade [Took a first pass](http://mattermost-dev-docs-preview-pulls.s3-website-us-east-1.amazonaws.com/1247/contribute/more-info/server/schema-migration-guide/) at updating global table styles and some custom edits for these specific tables. There were a few differences form the tables in the doc, but hopefully the colouring is okay 😅

**Comment by agnivade** [comment]
This looks amazing. Thanks @emdecr !

**Review by streamer45 — APPROVED** [review]
